efi/preinstall/check_pcr7.go

```diff
@@ -0,0 +1,529 @@
+					// secboot efi package rejects) or as a fallback for signed image where signature verification
+					// fails. Digests may be permitted for authenticating unsigned pre-OS components.
+				default:
+					// Make sure that we can parse all other signature databases ok
+					if _, err = efi.ReadSignatureDatabase(bytes.NewReader(data.VariableData)); err != nil {
+						return nil, fmt.Errorf("cannot decode %s contents from EV_EFI_VARIABLE_DRIVER_CONFIG event: %w", data.UnicodeName, err)
+					}
+				}
+			case tcglog.EventTypeEFIAction:
+				// This branch exists here for documentation purposes - it falls through to the
+				// default branch below, which returns an error.
+				//
+				// An EV_EFI_ACTION events with the string "UEFI Debug Mode" appears at the
+				// start of the log if a debugging endpoint is enabled. It's also possible that
+				// EV_EFI_ACTION events are used for other conditions in PCR7 that weaken device
+				// security (eg, the "DMA Protection Disabled" event).
+				//
+				// In general, it's not normal to see EV_EFI_ACTION events and these indicate some
+				// sort of abnormal condition that has a detrimental effect on device security.
+				// WithSecureBootPolicyProfile() will generate an invalid policy in this case because
+				// it doesn't emit them.
+				//
+				// Just return an error here to prevent the use of WithSecureBootPolicyProfile(). The
+				// "UEFI Debug Mode" and "DMA Protection Disabled" cases are already picked up by the
+				// firmware protection checks, so we don't need any special handling here.
+				fallthrough
+			default:
+				// Anything that isn't EV_EFI_VARIABLE_DRIVER_CONFIG ends up here.
+				return nil, fmt.Errorf("unexpected %v event %q whilst measuring config", ev.EventType, ev.Data)
+			}
+		case tcglogPhasePreOSAfterMeasureSecureBootConfig, tcglogPhaseOSPresent:
+			if len(configs) > 0 {
+				// We've transitioned to a phase where components can be loaded and verified but we haven't
+				// measured all of the secure boot variables. We'll fail to generate a valid policy with
+				// WithSecureBootPolicyProfile() in this case.
+				return nil, errors.New("EV_EFI_VARIABLE_DRIVER_CONFIG events for some secure boot variables missing from log")
+			}
+
+			if ev.PCRIndex == internal_efi.BootManagerCodePCR &&
+				ev.EventType == tcglog.EventTypeEFIBootServicesApplication &&
+				phase == tcglogPhaseOSPresent &&
+				!seenIBLLoadEvent {
+				// This is an EV_EFI_BOOT_SERVICES_APPLICATION event during OS-present,
+				// and we haven't seen the event for the IBL yet. We stop once we see this
+				// because at this point, the rest of the measurements in this PCR are under
+				// the control of the OS.
+				data, ok := ev.Data.(*tcglog.EFIImageLoadEvent)
+				if !ok {
+					return nil, fmt.Errorf("invalid OS-present EV_EFI_BOOT_SERVICES_APPLICATION event data: %w", ev.Data.(error))
+				}
+
+				yes, err := isLaunchedFromLoadOption(ev, bootOpt)
```

Comment:
there is similar code in check_pcr4, maybe we can extract a helper to make this a bit more compact/readable?